internal/controllers/variable_source.go

```diff
@@ -64,21 +63,35 @@ func (v *VariableSource) GetVariables(ctx context.Context) ([]deppy.Variable, er
 		return nil, err
 	}
 
-	// We are in process of getting rid of extra variable sources.
-	// See this for progress: https://github.com/operator-framework/operator-controller/issues/437
-	vs := variablesources.NestedVariableSource{
-		func(inputVariableSource input.VariableSource) (input.VariableSource, error) {
-			return variablesources.NewOperatorVariableSource(operatorList.Items, allBundles, inputVariableSource), nil
-		},
-		func(inputVariableSource input.VariableSource) (input.VariableSource, error) {
-			return variablesources.NewBundleDeploymentVariableSource(operatorList.Items, bundleDeploymentList.Items, allBundles, inputVariableSource), nil
-		},
-		func(inputVariableSource input.VariableSource) (input.VariableSource, error) {
-			return variablesources.NewBundlesAndDepsVariableSource(allBundles, inputVariableSource), nil
-		},
-		func(inputVariableSource input.VariableSource) (input.VariableSource, error) {
-			return variablesources.NewCRDUniquenessConstraintsVariableSource(inputVariableSource), nil
-		},
+	requiredPackages, err := variablesources.MakeRequiredPackageVariables(allBundles, operatorList.Items)
+	if err != nil {
+		return nil, err
+	}
+
+	installedPackages, err := variablesources.MakeInstalledPackageVariables(allBundles, operatorList.Items, bundleDeploymentList.Items)
+	if err != nil {
+		return nil, err
+	}
+
+	bundles, err := variablesources.MakeBundleVariables(allBundles, requiredPackages, installedPackages)
+	if err != nil {
+		return nil, err
+	}
+
+	bundleUniqueness := variablesources.MakeBundleUniquenessVariables(bundles)
+
+	result := []deppy.Variable{}
```

Comment:
Maybe take a look at https://freshman.tech/snippets/go/concatenate-slices/, specifically the section "Concatenating multiple slices at once", as a way to do this more efficiently?

Reply:
@ncdc the above works when we want to concatenate slices of the same type. Here we have slices of multiple different types all of which implement `deppy.Variable` interface and we need to create a slice of `deppy.Variable`s. I don't think it will work here.

I considered doing pre-allocation like this:

```go
result := make([]deppy.Variable, 0, len(requiredPackages)+len(installedPackages)+len(bundles)+len(bundleUniqueness))
```

But decided not to go with it. It still requires a loop with an `append`. Also It seems brittle: I can definitely see myself not updating the `make` call with a new variable when adding another loop. But it avoids extra allocations.